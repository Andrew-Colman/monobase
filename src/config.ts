import * as types from "./types";

type ProjectOptions = {
  path: string;
  build?: "debug" | "production";
  urlPrefix?: string;
};

export const project = ({
  path,
  build = "production",
  urlPrefix = ""
}: ProjectOptions): types.Project => {
  urlPrefix = urlPrefix.replace(/^\/+|\/+$/g, "");
  return {
    path: path,
    build: build,
    config: {
      pages: "pages",
      static: "static",
      components: "components",
      componentScript: "/components.js",
      extensions: ["js", "ts", "tsx", "mdx"],
      urlPrefix: urlPrefix.length > 0 ? `/${urlPrefix}` : ""
    }
  };
};
