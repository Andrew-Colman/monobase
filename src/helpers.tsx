import * as path from "path";
import * as React from "react";
import { renderToString } from "react-dom/server";
import * as styled from "styled-components";
import * as types from "./types";

// Component to inline styled component css
export const StyledSheet: React.SFC<{ app: React.ReactNode }> = props => {
  const sheet = new styled.ServerStyleSheet();

  renderToString(sheet.collectStyles(<>{props.app}</>));

  return (sheet.getStyleElement() as any) as React.ReactElement<any>;
};

export const Development = () => {
  const context = useContext();
  return <script src={relative(context.project.config.componentScript)} />;
};

export const useContext = (): types.Context => {
  if (!process.env["context"]) {
    throw Error(
      "process.env.context is missing. You might be using useContext in a Dynamic component? "
    );
  }
  // This gets inserted by webpack on build
  return (process.env["context"] as any) as types.Context;
};

export const useProject = (): types.Project => {
  return useContext().project;
};

export const relative = (to: string) => {
  const context = useContext();
  return path.relative(context.url, to);
};
