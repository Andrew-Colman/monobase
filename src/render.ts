import * as path from "path";
import * as fs from "fs";
import { renderToString } from "react-dom/server";
import { Compiler, Config } from "./compiler";
import * as types from "./types";
import * as dynamic from "./dynamic";
import * as context from "./context";
import { memoize } from "lodash";

// We memoize the script compiler based on the config for fast reloads
// as long as the dynamic components have not changed on disk.

const getCachedCompiler = memoize(({ name, projectPath, config }) => {
  return new Compiler(projectPath, config);
}, JSON.stringify);

export const page = async (project: types.Project, page: string) => {
  const pagePath = path.join(project.config.pages, page);

  // const compiler = new Compiler(config as any);
  const compiler = getCachedCompiler({
    name: "page",
    projectPath: project.path,
    config: {
      cache: true,
      externals: true
    }
  });

  // A syntax error could occur here
  await compiler.compile([pagePath], context.create(project, ""));

  // Temporary write the generated javascript for this page for debug purposes
  // const pageScriptPath = path.join(project.path, "build", page + ".js");
  // fs.writeFileSync(pageScriptPath, compiler._output);

  if (!compiler.module["default"]) {
    throw Error(
      `Missing default export for page ${page}. Did you maybe forget to add "export default"`
    );
  }

  // An eval runtime could happen here
  const pageModule = compiler.module["default"](project);
  const html = renderToString(pageModule);

  // Clean up the generated javascript file
  // fs.unlinkSync(pageScriptPath);

  return html;
};

export const script = async (project: types.Project) => {
  const entries = dynamic.entries(project);

  const compiler = getCachedCompiler({
    name: "script",
    projectPath: project.path,
    config: {
      production: project.build === "production",
      cache: true,
      externals: false
    }
  });

  return compiler.compile(entries, {});
};
