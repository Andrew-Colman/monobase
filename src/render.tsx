import * as path from "path";
import * as fs from "fs";
import { renderToString } from "react-dom/server";
import { Compiler, Config } from "./compiler";
import * as types from "./types";
import * as dynamic from "./dynamic";
import * as context from "./context";
import { memoize } from "lodash";
import * as React from "react";
import * as prettyBytes from "pretty-bytes";
import { urlForPage } from "./resolve";
import { PageContext, PageContextType } from "./contexts";

// We memoize the script compiler based on the config for fast reloads
// as long as the dynamic components have not changed on disk.

const getCachedCompiler = memoize(({ name, projectPath, config }) => {
  return new Compiler(projectPath, config);
}, JSON.stringify);

export const page = async (project: types.Project, page: string) => {
  const pagePath = path.join(project.config.pages, page);

  // const compiler = new Compiler(config as any);
  const compiler = getCachedCompiler({
    name: "page",
    projectPath: project.path,
    config: {
      cache: true,
      externals: true
    }
  });

  // A syntax error could occur here
  await compiler.compile([pagePath], context.create(project, pagePath));
  // This is definitely not cool, but some caching has me run this twice
  await compiler.compile([pagePath], context.create(project, pagePath));

  const compilerModule = compiler.module;

  // Temporary write the generated javascript for this page for debug purposes
  // const pageScriptPath = path.join(project.path, "build", page + ".js");
  // fs.writeFileSync(pageScriptPath, compiler._output);

  if (!compilerModule["default"]) {
    throw Error(
      `Missing default export for page ${page}. Did you maybe forget to add "export default"`
    );
  }

  const pageContext: PageContextType = {
    project,
    path: pagePath,
    styles: compiler.styles
  };

  const PageModule = compilerModule["default"];

  const Page = function() {
    return (
      <PageContext.Provider value={pageContext}>
        <PageModule />
      </PageContext.Provider>
    );
  };

  const html = renderToString(React.createElement(Page));

  // Clean up the generated javascript file
  // fs.unlinkSync(pageScriptPath);

  return html;
};

export const script = async (project: types.Project) => {
  const entries = dynamic.entries(project);

  const compiler = getCachedCompiler({
    name: "script",
    projectPath: project.path,
    config: {
      production: project.build === "production",
      cache: true,
      externals: false
    }
  });

  return compiler.compile(entries, {});
};

// export const styles = async (project: types.Project) => {
//   const entries = dynamic.entries(project);

//   const compiler = getCachedCompiler({
//     name: "script",
//     projectPath: project.path,
//     config: {
//       production: project.build === "production",
//       cache: true,
//       externals: false
//     }
//   });

//   const result = await compiler.compile(entries, {});

//   console.log("compiler.styles", compiler.styles);

//   return compiler.styles;
// };
