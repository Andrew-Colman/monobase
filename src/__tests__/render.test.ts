import { __setContextForTest } from "../helpers";
import { project } from "../config";
import * as render from "../render";
import { resolve } from "path";

// https://github.com/styled-components/styled-components/issues/1692
import * as styled from "styled-components";

const { StyleSheet } = styled[
  "__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS"
];

describe("output", () => {
  beforeEach(() => {
    StyleSheet.reset(true);
    __setContextForTest({
      path: "pages",
      project: project({ path: "pages" }),
      url: "/about"
    });
  });

  afterEach(() => __setContextForTest(null));

  const proj = project({ path: resolve(__dirname, "../../project") });

  test("/index.tsx", async () => {
    expect(await render.page(proj, "/index.tsx")).toMatchSnapshot();
  });

  test("/about.tsx", async () => {
    expect(await render.page(proj, "/about.tsx")).toMatchSnapshot();
  });

  test("/article.mdx", async () => {
    expect(await render.page(proj, "/article.mdx")).toMatchSnapshot();
  });
});
