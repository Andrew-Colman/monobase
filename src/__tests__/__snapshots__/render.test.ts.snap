// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output /about.tsx 1`] = `
"<html data-reactroot=\\"\\"><head><meta charSet=\\"utf-8\\"/><link rel=\\"stylesheet\\" href=\\"/static/styles.css\\"/></head><body><div style=\\"font:11px/1.6em Menlo\\"><h2>Context</h2><pre style=\\"font:11px/1.6em Menlo\\">{
    &quot;project&quot;: {
        &quot;path&quot;: &quot;/Users/koen/Documents/Projects/monobase/project&quot;,
        &quot;build&quot;: &quot;production&quot;,
        &quot;config&quot;: {
            &quot;pages&quot;: &quot;pages&quot;,
            &quot;static&quot;: &quot;static&quot;,
            &quot;components&quot;: &quot;components&quot;,
            &quot;componentScript&quot;: &quot;/components.js&quot;,
            &quot;extensions&quot;: [
                &quot;js&quot;,
                &quot;ts&quot;,
                &quot;tsx&quot;,
                &quot;mdx&quot;
            ],
            &quot;urlPrefix&quot;: &quot;&quot;
        }
    },
    &quot;path&quot;: &quot;pages/about.tsx&quot;,
    &quot;url&quot;: &quot;/about/&quot;
}</pre></div><script src=\\"../components.js\\"></script></body></html>"
`;

exports[`output /article.mdx 1`] = `
"<html data-reactroot=\\"\\"><head><meta charSet=\\"utf-8\\"/><link rel=\\"stylesheet\\" href=\\"/static/styles.css\\"/></head><body><h2>Hello Koen</h2><p>This is a paragraph.</p><component data-component-props=\\"[&quot;Component.Button&quot;,{},[]]\\"><button style=\\"padding:24px 50px 26px;user-select:none;border-radius:8px;border:0;outline:0;line-height:1;font-size:24px;background:#0AF;color:#fff\\">Count: <!-- -->0</button></component><pre><code>Hello world
</code></pre><script src=\\"../components.js\\"></script></body></html>"
`;

exports[`output /index.tsx 1`] = `
"<html data-reactroot=\\"\\"><head><meta charSet=\\"utf-8\\"/><style data-styled=\\"eQEdJi gGJNfi itrzlT jwMcgt\\" data-styled-version=\\"4.3.2\\">
/* sc-component-id: Styled__Button-sc-1yi0x62-0 */
.gGJNfi{font-size:1em;margin:1em;padding:0.25em 1em;border-radius:3px;color:royalblue;border:2px solid royalblue;}.itrzlT{font-size:1em;margin:1em;padding:0.25em 1em;border-radius:3px;color:mediumseagreen;border:2px solid mediumseagreen;}.jwMcgt{font-size:1em;margin:1em;padding:0.25em 1em;border-radius:3px;color:darkorange;border:2px solid darkorange;}
/* sc-component-id: Styled__ButtonB-sc-1yi0x62-1 */
.eQEdJi{background:#09F;border-radius:4px;border:none;color:#fff;font-size:1em;margin:1em;outline:0;padding:0.25em 1em;}</style><link rel=\\"stylesheet\\" href=\\"/static/styles.css\\"/></head><body><span style=\\"text-align:center\\"><section style=\\"padding:60px\\"><h1>Welcome to Monobase</h1><p>A simple React based static site generator</p></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Button</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Button&quot;,{},[]]\\"><button style=\\"padding:24px 50px 26px;user-select:none;border-radius:8px;border:0;outline:0;line-height:1;font-size:24px;background:#0AF;color:#fff\\">Count: <!-- -->0</button></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Mouse Location</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Mouse&quot;,{},[]]\\"><button style=\\"padding:24px 50px 26px;user-select:none;border-radius:8px;border:1px solid #F0F0F0;outline:0;line-height:1;font-size:24px;font-family:monospace;background:white\\">0<!-- -->, <!-- -->0</button></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Time</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Timer&quot;,{},[]]\\"><button style=\\"padding:24px 50px 26px;user-select:none;border-radius:8px;border:1px solid #F0F0F0;outline:0;line-height:1;font-size:24px;font-family:monospace\\">0</button></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">CSS Styled Component</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Styled&quot;,{},[]]\\"><div><div><button class=\\"Styled__ButtonB-sc-1yi0x62-1 eQEdJi\\">Click Error</button><button class=\\"Styled__Button-sc-1yi0x62-0 gGJNfi\\">Ad hoc theme</button><div><button class=\\"Styled__Button-sc-1yi0x62-0 itrzlT\\">Themed</button><button class=\\"Styled__Button-sc-1yi0x62-0 jwMcgt\\">Overidden</button></div></div></div></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Visible</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Visible&quot;,{&quot;inset&quot;:100},&quot;I&#x27;m like a refrigerator light. Or Schrödingers cat.&quot;]\\"><div style=\\"visibility:visible\\">I&#x27;m like a refrigerator light. Or Schrödingers cat.</div></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Enter a color name</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Colors&quot;,{},[]]\\"><div style=\\"width:300px;margin:auto\\"><div style=\\"height:300px;border-radius:8px 8px 0 0;transition:background-color .2s;background-color:#EEE\\"></div><input type=\\"text\\" value=\\"\\" placeholder=\\"like &#x27;darksalmon&#x27;\\" style=\\"padding:24px 50px 26px;user-select:none;border-radius:0 0 8px 8px;border:1px solid #EEE;outline:0;line-height:1;font-size:24px;width:100%;border-top:none;text-align:center\\"/></div></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Persistent Cookie</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.Cookie&quot;,{},[]]\\"><input type=\\"text\\" value=\\"\\" placeholder=\\"Type something…\\" style=\\"padding:24px 50px 26px;user-select:none;border-radius:8px;border:1px solid #EEE;outline:0;line-height:1;font-size:24px;width:100%;text-align:center\\"/></component></div></section><section style=\\"text-align:center;padding:80px 10px\\"><h3 style=\\"font-size:30px;font-weight:700;margin:0;padding-bottom:40px;line-height:1\\">Contexts</h3><div style=\\"text-align:center\\"><component data-component-props=\\"[&quot;Component.ContextExample&quot;,{},[]]\\"><button>Update</button><div>Count: <!-- -->0</div><div>Count: <!-- -->0</div><div>Count: <!-- -->0</div></component></div></section></span><script src=\\"components.js\\"></script></body></html>"
`;
